File: backend/meta/property.py
"""
属性实体
Property必须属于某个Model
"""
from typing import Optional, List, Any
import json


class Property:
    """属性实体 - 必须属于Model"""
    
    def __init__(
        self,
        id: int,
        name: str,
        code: str,
        type: str,
        modelId: int,
        required: bool = False,
        description: Optional[str] = None,
        isPrimaryKey: bool = False,
        isForeignKey: bool = False,
        defaultValue: Optional[str] = None,
        constraints: Optional[List[Any]] = None,
        sensitivityLevel: Optional[str] = None,
        maskRule: Optional[str] = None,
        physicalColumn: Optional[str] = None,
        foreignKeyTable: Optional[str] = None,
        foreignKeyColumn: Optional[str] = None
    ):
        self.id = id
        self.name = name
        self.code = code
        self.type = type
        self.modelId = modelId
        self.required = required
        self.description = description
        self.isPrimaryKey = isPrimaryKey
        self.isForeignKey = isForeignKey
        self.defaultValue = defaultValue
        self.constraints = constraints or []
        self.sensitivityLevel = sensitivityLevel
        self.maskRule = maskRule
        self.physicalColumn = physicalColumn
        self.foreignKeyTable = foreignKeyTable
        self.foreignKeyColumn = foreignKeyColumn
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "code": self.code,
            "type": self.type,
            "required": self.required,
            "description": self.description,
            "modelId": self.modelId,
            "isPrimaryKey": self.isPrimaryKey,
            "isForeignKey": self.isForeignKey,
            "defaultValue": self.defaultValue,
            "constraints": self.constraints,
            "sensitivityLevel": self.sensitivityLevel,
            "maskRule": self.maskRule,
            "physicalColumn": self.physicalColumn,
            "foreignKeyTable": self.foreignKeyTable,
            "foreignKeyColumn": self.foreignKeyColumn
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Property':
        """从字典创建实例"""
        constraints = data.get("constraints", [])
        if isinstance(constraints, str):
            try:
                constraints = json.loads(constraints)
            except json.JSONDecodeError:
                constraints = []
        
        return cls(
            id=data["id"],
            name=data["name"],
            code=data["code"],
            type=data["type"],
            modelId=data["modelId"],
            required=data.get("required", False),
            description=data.get("description"),
            isPrimaryKey=data.get("isPrimaryKey", False),
            isForeignKey=data.get("isForeignKey", False),
            defaultValue=data.get("defaultValue"),
            constraints=constraints,
            sensitivityLevel=data.get("sensitivityLevel"),
            maskRule=data.get("maskRule"),
            physicalColumn=data.get("physicalColumn"),
            foreignKeyTable=data.get("foreignKeyTable"),
            foreignKeyColumn=data.get("foreignKeyColumn")
        )
    
    def __repr__(self):
        return f"Property(id={self.id}, name='{self.name}', modelId={self.modelId})"

File: backend_java/src/main/java/com/jianmo/platform/entity/Property.java
package com.jianmo.platform.entity;

import javax.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.ToString;

import java.time.LocalDateTime;


@Entity
@Table(name = "properties")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Property {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String code;

    @Column(nullable = false)
    private String type;

    @Column(name = "model_id")
    private Long modelId;

    @Column(name = "is_required")
    private Boolean isRequired = false;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "is_primary_key")
    private Boolean isPrimaryKey = false;

    @Column(name = "is_foreign_key")
    private Boolean isForeignKey = false;

    @Column(name = "default_value")
    private String defaultValue;

    @Column(columnDefinition = "TEXT")
    private String constraints;

    @Column(name = "sensitivity_level")
    private String sensitivityLevel;

    @Column(name = "mask_rule")
    private String maskRule;

    @Column(name = "physical_column")
    private String physicalColumn;

    @Column(name = "foreign_key_table")
    private String foreignKeyTable;

    @Column(name = "foreign_key_column")
    private String foreignKeyColumn;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "model_id", insertable = false, updatable = false)
    @ToString.Exclude
    private Model model;

    
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public Property(String name, String code, String type, Long modelId) {
        this.name = name;
        this.code = code;
        this.type = type;
        this.modelId = modelId;
    }

    public boolean isValid() {
        if (code == null || code.isEmpty()) {
            return false;
        }
        return code.matches("^[a-zA-Z][a-zA-Z0-9_]*$");
    }
}
